File: M12 NF1 (Jose Java)/Softlearning/model/Products/Dimensions.php
<?php




class Dimensions {
    protected float $weight;
    protected float $height;
    protected float $width;
    protected float $length;
    protected float $volume;
    protected bool $create = true;

    public function __construct(float $weight, float $height, float $width, float $length) {
        $errors = '';

        if ($this->setWeight($weight) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($weight)) . "\n";
        }

        if ($this->setHeight($height) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($height)) . "\n";
        }

        if ($this->setWidth($width) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($width)) . "\n";
        }

        if ($this->setLength($length) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($length)) . "\n";
        }

        if (!empty($errors)) {
            throw new BuildException("Not possible to create the dimensions: \n" . $errors);
        }
    }

    public function getWeight(): float {
        return $this->weight;
    }

    public function getHeight(): float {
        return $this->height;
    }

    public function getWidth(): float {
        return $this->width;
    }

    public function getLength(): float {
        return $this->length;
    }

    public function setWeight(float $weight): int {
        $errorWeight = Check::validate($weight);
        if ($errorWeight == 0) {
            $this->weight = $weight;
        }
        return $errorWeight;
    }

    public function setHeight(float $height): int {
        $errorHeight = Check::validate($height);
        if ($errorHeight == 0) {
            $this->height = $height;
        }
        return $errorHeight;
    }

    public function setWidth(float $width): int {
        $errorWidth = Check::validate($width);
        if ($errorWidth == 0) {
            $this->width = $width;
        }
        return $errorWidth;
    }

    public function setLength(float $length): int {
        $errorLength = Check::validate($length);
        if ($errorLength == 0) {
            $this->length = $length;
        }
        return $errorLength;
    }

    public function getVolume(): float {
        return $this->width * $this->height * $this->length;
    }

    public function getDetails(): string {
        return "Height: " . $this->getHeight() . " cm\n" .
               "Weight: " . $this->getWeight() . " kg\n" .
               "Width: " . $this->getWidth() . " cm\n" .
               "Length: " . $this->getLength() . " cm\n" .
               "Volume: " . $this->getVolume() . " cubic cm";
    }
}











































/*
class Dimensions {
    protected float $weight;
    protected float $height;
    protected float $width;
    protected float $length;
    protected float $volume;
    protected bool $create = true;

    public function __construct(float $weight, float $height, float $width, float $length) {
        if ($this->setWeight($weight) !== 0) {
            $this->create = false;
        }

        if ($this->setHeigth($height) !== 0) {
            $this->create = false;
        }

        if ($this->setWidth($width) !== 0) {
            $this->create = false;
        }

        if ($this->setLength($length) !== 0) {
            $this->create = false;
        }

        if (!$this->create) {
            throw new BuildException("Not possible to create the dimensions.");
        }
    }

    public function getWeight(): float {
        return $this->weight;
    }

    public function getHeight(): float {
        return $this->heigth;
    }

    public function getWidth(): float {
        return $this->width;
    }

    public function getLength(): float {
        return $this->length;
    }

    public function setWeight(float $weight): int {
        $errorWeight = Check::validate($weight);
        if ($errorWeight != 0) {
            print Check::getErrorMessage($errorWeight);
        } else {
            $this->weight = $weight;
            print "<br> El peso de este libro es: " . $weight;
        }
        return $errorWeight;
    }

    public function setheight(float $heigth): int {
        $errorHeigth = Check::validate($heigth);
        if ($errorHeigth != 0) {
            print Check::getErrorMessage($errorHeigth);
        } else {
            $this->heigth = $heigth;
            ;
        }
        return $errorHeigth;
    }

    public function setWidth(float $width): int {
        $errorWidth = Check::validate($width);
        if ($errorWidth != 0) {
            print Check::getErrorMessage($errorWidth);
        } else {
            $this->width = $width;
            print "<br>El ancho de este libro es: " . $width;
        }
        return $errorWidth;
    }

   public function setLength(float $length): int {
        $errorLength = Check::validate($length);
        if ($errorLength != 0) {
            print Check::getErrorMessage($errorLength);
        } else {
            $this->length = $length;
            print "<br>El largo de este libro es: " . $length;
        }
        return $errorLength;
    }

    public function getVolume(): float {
        return $this->width * $this->heigth * $this->length;
    }

    public function getDetails(): string {
        return "Heigth: " . $this->getHeigth() . ", Weight: " . $this->getWeight() . ", Width: " . $this->getWidth() . ", Length: " . $this->getLength();
    }
}
*/


/*
class Dimensions {

    protected float $weight;
    protected float $heigth;
    protected float $width;
    protected float $length;
    protected float $volume;
    protected bool $create=true;

    public function __construct(float $weight, float $heigth, float $width, float $length) {

        $errorWeight = Check::validate($weight);
        if ($errorWeight != 0) {
            print Check::getErrorMessage($errorWeight);
            $this->create = false;
        } else {
            $this->setWeight($weight);
            print "<br> El peso de este libro es: " . $weight;
        }

        $errorHeigth = Check::validate($heigth);
        if ($errorHeigth == 0) {
            $this->setHeigth($heigth);
            print "<br>La altura de este libro es: " . $heigth;
        } else {
            print Check::getErrorMessage($errorHeigth);
            $this->create = false;
        }
        $errorWidth = Check::validate($width);
        if ($errorWidth == 0) {
            $this->setWidth($width);
            print "<br>El ancho de este libro es: " . $width;
        } else {
            print Check::getErrorMessage($errorWidth);
            $this->create = false;
        }

        $errorLength = Check::validate($length);
        if ($errorLength == 0) {

            print "<br>El largo de este libro es: " . $length;
        } else {
            $this->create = false;
            print Check::getErrorMessage($errorLength);
        }
    }

    public function getWeight(): string {
        return $this->weight;
    }

    public function getHeigth(): string {
        return $this->heigth;
    }

    public function getWidth(): string {
        return $this->width;
    }

    public function setWeight(float $weight): void {
        $this->weight = $weight;
    }

    public function setHeigth(float $heigth): void {
        $this->heigth = $heigth;
    }

    public function setWidth(float $width): void {
        $this->width = $width;
    }

    public function getVolume(): float {
        $volume = $this->width * $this->heigth * $this->weight;
        return $this->volume = $this->width * $this->heigth * $this->weight;
    }

    public function getDetails() {
        return $this->getHeigth() . $this->getWeight() . $this->getWidth();
    }
}
*/

File: src/main/java/com/core/entities/shared/dimensions/Dimensions.java
// Version: 1.0
package com.core.entities.shared.dimensions; 
import com.core.checks.Check;
import com.core.entities.exceptions.BuildException;


public class Dimensions {
    protected double weight;
    protected double height;
    protected double width;
    protected boolean fragile;
    protected double length;
    protected double volume;
    
    protected Dimensions(){

    };
    public static Dimensions getInstanceDimensions(double weight, double height, double width, boolean fragile, double length) throws BuildException{ 
        StringBuilder errors = new StringBuilder();

        
        Dimensions d = new Dimensions();
        int errorCode;


        if ((errorCode = d.setWeight(weight)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        if ((errorCode = d.setHeight(height)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        if ((errorCode = d.setWidth(width)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }
        if ((errorCode = d.setFragile(fragile)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        if ((errorCode = d.setLength(length)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        if (errors.length() > 0) {
            
            throw new BuildException("Not possible to create the dimensions: \n" + errors.toString());
        }
        return d;
    }

    public double  getWeight() {
        return this.weight;
    }

    public double getHeight() {
        return this.height;
    }

    public double getWidth() {
        return this.width;
    }

    public boolean getFragile() {
        return this.fragile;
    }

    public double getLength() {
        return this.length;
    }
    
    public double getVolume() {
        return this.width * this.height * this.length;
    }

    // public String getVolumeDetails() {
    //     // es un bloque de texto, como si escribieras en un note pad, se usa para escribir un string largo sin tener que concatenar
    //     // preguntar por que no hace falta cerrarlo
    //     return """
    //             Height: """ + this.getHeight() + " cm\n" +
    //             "Weight: " + this.getWeight() + " kg\n" +
    //             "Width: " + this.getWidth() + " cm\n" +
    //             "Fragile: " + this.getFragile() + "\n" +
    //             "Length: " + this.getLength() + " cm\n" +
    //             "Volume: " + this.getVolume() + " cubic cm";
    // }
    
    //agregar un metodo toString
    public String toString() {
        return "Height: " + this.getHeight() + " cm\n" +
                "Weight: " + this.getWeight() + " kg\n" +
                "Width: " + this.getWidth() + " cm\n" +
                "Fragile: " + this.getFragile() + "\n" +
                "Length: " + this.getLength() + " cm\n" +
                "Volume: " + this.getVolume() + " cubic cm";
    }

    // @Override
    // public StringBuilder toString() {
    //     StringBuilder details = new StringBuilder();
    //     details.append("Height: ").append(this.getHeight()).append(" cm\n")
    //            .append("Weight: ").append(this.getWeight()).append(" kg\n")
    //            .append("Width: ").append(this.getWidth()).append(" cm\n")
    //            .append("Fragile: ").append(this.getFragile()).append("\n")
    //            .append("Length: ").append(this.getLength()).append(" cm\n")
    //            .append("Volume: ").append(this.getVolume()).append(" cubic cm");
    //     return details;
    // }
    


    

    public int setWeight(double weight) {
        int errorWeight = Check.range(weight);
        if (errorWeight == 0) {
            this.weight = weight;
        }
        return errorWeight;
    }

    public int setHeight(double height) {
        int errorHeight = Check.range(height);
        if (errorHeight == 0) {
            this.height = height;
        }
        return errorHeight;
    }

    public int setWidth(double width) {
        int errorWidth = Check.range(width);
        if (errorWidth == 0) {
            this.width = width;
        }
        return errorWidth;
    }

    public int setLength(double length) {
        int errorLength = Check.range(length);
        if (errorLength == 0) {
            this.length = length;
        }
        return errorLength;
    }

    public int setFragile(boolean f) {
        if(f == true || f == false){
            this.fragile = f;
            return 0;
        }
        return -18;
    }

}